src/System.Management.Automation/engine/runtime/ScriptBlockToPowerShell.cs

```diff
@@ -413,34 +478,10 @@ private ScriptBlockToPowerShellConverter()
             while (currentParent != null)
```

Comment:
You can avoid `foreachNestedCount` and `CheckScriptBlockForForeachCommand` if you make this check to be `while (currentParent != scriptBlock.Ast)`.

That while-check means the inspection goes up the AST chain until it reaches at the `ScriptBlockAst` that is going to be executed, so as soon as you find one `ForEach-Object -Parallel` during this walkup, you will know the `usingAst` is in a nested `ForEach-Object -Parallel` and can return false.